pojo/src/main/java/com/spotify/hamcrest/pojo/IsPojo.java

```diff
@@ -50,20 +60,66 @@ private IsPojo(final Class<A> cls, final ImmutableMap<String, Matcher<?>> method
   }
 
   public IsPojo<A> where(String methodName, Matcher<?> returnValueMatcher) {
-    final ImmutableMap<String, Matcher<?>> newMethodMatchers =
-        ImmutableMap.<String, Matcher<?>>builder()
-            .putAll(methodMatchers)
-            .put(methodName, returnValueMatcher)
-            .build();
+    return whereWithoutTypeSafety(new MethodValueSupplier<>(methodName), returnValueMatcher);
+  }
 
-    return new IsPojo<>(cls, newMethodMatchers);
+  public <T> IsPojo<A> where(ValueProvider<A, T> valueProvider, Matcher<T> valueMatcher) {
+    return whereWithoutTypeSafety(
+        new NamedLambdaValueSupplier<>(
+            extractLambdaName(valueProvider),
+            valueProvider),
+        valueMatcher);
   }
 
   public IsPojo<A> withProperty(String property, Matcher<?> valueMatcher) {
     return where("get" + CaseFormat.LOWER_CAMEL.to(CaseFormat.UPPER_CAMEL, property),
                  valueMatcher);
   }
 
+  private IsPojo<A> whereWithoutTypeSafety(Function<A, ?> valueSupplier, Matcher<?> valueMatcher) {
+    return new IsPojo<>(cls,
+                        ImmutableMap.<Function<A, ?>, Matcher<?>>builder()
+                            .putAll(methodMatchers)
+                            .put(valueSupplier, valueMatcher)
+                            .build());
+  }
+
+  /**
+   * Method uses serialization trick to extract information about lambda,
+   * to give understandable name in case of mismatch.
+   * @param lambda lambda to extract the name from
+   * @return string describing class and method from which lambda was created,
+   *     or simple {@code toString()} if that fails.
+   */
+  private String extractLambdaName(Object lambda) {
```

Comment:
I think this should be extracted as a static function somewhere else. Either in a helper class or in the `Lambda` class if needed